Simplenote/src/main/java/com/automattic/simplenote/PreferencesFragment.java

```diff
@@ -456,6 +484,109 @@ public int compare(String text1, String text2) {
         }
     }
 
+    private void importData(Uri uri) {
+        Simplenote currentApp = (Simplenote) requireActivity().getApplication();
+        Bucket<Note> noteBucket = currentApp.getNotesBucket();
+
+        try {
+            String noteContent = readFile(uri);
+            if ("json".equals(getFileExtension(uri))) {
+                JSONObject jsonData = new JSONObject(noteContent);
+                JSONArray activeNotes = jsonData.getJSONArray("activeNotes");
+                JSONArray trashedNotes = jsonData.getJSONArray("trashedNotes");
+                List<String> trashedIds = new ArrayList<>();
+                JSONArray notesArray = new JSONArray();
+                for (int i = 0; i < activeNotes.length(); i++) {
+                    notesArray.put(activeNotes.getJSONObject(i));
+                }
+                for (int j = 0; j < trashedNotes.length(); j++) {
+                    notesArray.put(trashedNotes.getJSONObject(j));
+                    trashedIds.add(trashedNotes.getJSONObject(j).getString("id"));
+                }
+                for (int k = 0; k < notesArray.length(); k++) {
+                    JSONObject noteJson = notesArray.getJSONObject(k);
+                    Note note = noteBucket.newObject();
+                    Iterator<String> keys = noteJson.keys();
+                    while (keys.hasNext()) {
+                        String key = keys.next();
+                        switch (key) {
+                            case "id":
+                                note.setDeleted(trashedIds.contains(noteJson.getString(key)));
+                                break;
+                            case "content":
+                                note.setContent(noteJson.getString(key));
+                                break;
+                            case "creationDate":
+                                note.setCreationDate(DateTimeUtils.getDateCalendar(noteJson.getString(key)));
+                                break;
+                            case "lastModified":
+                                note.setModificationDate(DateTimeUtils.getDateCalendar(noteJson.getString(key)));
+                                break;
+                            case "tags":
+                                note.setTags(jsonToList(noteJson.getJSONArray(key)));
+                                break;
+                            case "pinned":
+                                note.setPinned(noteJson.getBoolean(key));
+                                break;
+                            case "markdown":
+                                note.setMarkdownEnabled(noteJson.getBoolean(key));
+                                break;
+                            case "publicURL":
+                                // note.setPublished(?);
+                                break;
+                        }
+                    }
```

Comment:
the note data structure has a relatively fixed form. do we need to iterate over all the keys? I think we'll want all these fields and directly assigning them would be clearer and terser. that is, we don't really want to use a note if it doesn't contain `content` but this iteration+switch is happy to do just that.

```java
note.setContent(noteJson.has("content") ? noteJson.getString("content") : "");
note.setCreationDate(noteJson.has("creationDate") ? DateTimeUtils.getDateCalendar(noteJson.getString("creationDate")) : now());
```

could be nice to move this into a new function such as `noteFromJson(JSONValue json, List<String> trashedIds)` and then it'd be easier to test in isolation as well.

Reply:
First wrote it like this, but then found a solution with an iterator better. I'll change it! And where should I place the `noteFromJson(JSONValue json, List<String> trashedIds)` method? Ok if I leave it in a PreferencesFragment?

Reply:
Good question. You might play around with it some and tell us what you think. I could see it having a good home in `Note` as well. That would bring it closer to the rest of the note semantics/models.

Reply:
Good, thank you, I will think about it.